File: bk2ch23p824peopleGroupsCoreData/ch36p1079peopleGroupsCoreData/PeopleLister.m


#import "PeopleLister.h"
#import "NSManagedObject+GroupAndPerson.h"

@interface PeopleLister () <NSFetchedResultsControllerDelegate, UITableViewDelegate, UITableViewDataSource, UITextFieldDelegate>

@property (nonatomic, strong) NSManagedObject* groupObject;
@property (nonatomic, strong) NSFetchedResultsController* frc;

@end

@implementation PeopleLister

- (id) initWithNibName:(NSString *)nibNameOrNil 
                bundle:(NSBundle *)nibBundleOrNil 
               groupManagedObject: (NSManagedObject*) object {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self->_groupObject = object;
    }
    return self;
}

- (id) initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    return nil; // we must not exist without a group object
}


- (void) viewDidLoad {
    [super viewDidLoad];
    
    self.title = self.groupObject.name;
    UIBarButtonItem* b = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(doAdd:)];
    self.navigationItem.rightBarButtonItems = @[b];
    
    [self.tableView registerNib:[UINib nibWithNibName:@"PersonCell" bundle:nil] forCellReuseIdentifier:@"Person"];
}

- (NSFetchedResultsController *)frc {
    if (_frc != nil) {
        return _frc;
    }
    
    NSFetchRequest *req = [[NSFetchRequest alloc] init];
    NSEntityDescription *entity =
    [NSEntityDescription entityForName:@"Person"
                inManagedObjectContext:self.groupObject.managedObjectContext];
    req.entity = entity;
    req.fetchBatchSize = 20;
    NSSortDescriptor *sortDescriptor =
    [[NSSortDescriptor alloc] initWithKey:@"timestamp" ascending:YES];
    req.sortDescriptors = @[sortDescriptor];
    NSPredicate* pred = [NSPredicate predicateWithFormat:@"group = %@", self.groupObject];
    req.predicate = pred;
    
    NSFetchedResultsController *afrc =
    [[NSFetchedResultsController alloc]
     initWithFetchRequest:req
     managedObjectContext:self.groupObject.managedObjectContext
     sectionNameKeyPath:nil
     cacheName:self.groupObject.uuid]; // prevent cache name conflicts
    afrc.delegate = self;
    self.frc = afrc;
    
	NSError *error = nil;
	if (![self.frc performFetch:&error]) {
        // Replace this implementation with code to handle the error appropriately.
        // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
	    NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
	    abort();
	}
    
    return _frc;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [[self.frc sections] count];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    id<NSFetchedResultsSectionInfo> sectionInfo = [self.frc sections][section];
    return [sectionInfo numberOfObjects];
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Person" forIndexPath:indexPath];
    NSManagedObject *object = [self.frc objectAtIndexPath:indexPath];
    UITextField* first = (UITextField*)[cell viewWithTag:1];
    UITextField* last = (UITextField*)[cell viewWithTag:2];
    first.text = object.firstName;
    last.text = object.lastName;
    first.delegate = last.delegate = self;
    return cell;
}

- (void) doAdd: (id) sender {
    [self.tableView endEditing:YES];
    NSManagedObjectContext *context = self.frc.managedObjectContext;
    NSEntityDescription *entity = self.frc.fetchRequest.entity;
    NSManagedObject *mo =
    [NSEntityDescription insertNewObjectForEntityForName:[entity name] inManagedObjectContext:context];
    mo.group = self.groupObject;
    mo.lastName = @"";
    mo.firstName = @"";
    mo.timestamp = [NSDate date];
    // save context
    NSError *error = nil;
    BOOL ok = [context save:&error];
    if (!ok) {
        NSLog(@"%@", error);
        return;
    }
    // and the rest is in the update delegate messages
}

-(void)textFieldDidEndEditing:(UITextField *)textField {
    // NSLog(@"did end");
    UIView* v = textField.superview;
    while (![v isKindOfClass: [UITableViewCell class]])
        v = v.superview;
    UITableViewCell* cell = (UITableViewCell*)v;
    NSIndexPath* ip = [self.tableView indexPathForCell:cell];
    NSManagedObject* object = [self.frc objectAtIndexPath:ip];
    [object setValue:textField.text forKey: ((textField.tag == 1) ? @"firstName" : @"lastName")];
    
    // save context
    NSError *error = nil;
    BOOL ok = [object.managedObjectContext save:&error];
    if (!ok) {
        NSLog(@"%@", error);
        return;
    }
}

-(void) viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    NSManagedObjectContext *context = self.frc.managedObjectContext;
    // save context
    NSError *error = nil;
    BOOL ok = [context save:&error];
    if (!ok) {
        NSLog(@"%@", error);
        return;
    }
}

#pragma mark - Content Update

-(void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView beginUpdates];
}

-(void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView endUpdates];
}

-(void)controller:(NSFetchedResultsController *)controller
  didChangeObject:(id)anObject
      atIndexPath:(NSIndexPath *)indexPath
    forChangeType:(NSFetchedResultsChangeType)type
     newIndexPath:(NSIndexPath *)newIndexPath {
    if (type == NSFetchedResultsChangeInsert) {
        [self.tableView insertRowsAtIndexPaths:@[newIndexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        dispatch_async(dispatch_get_main_queue(), ^{ // wait for interface to settle
            UITableViewCell* cell = [self.tableView cellForRowAtIndexPath:newIndexPath];
            UITextField* tf = (UITextField*)[cell viewWithTag:1];
            [tf becomeFirstResponder];
        });
    }
}


@end

File: bk2ch23p824peopleGroupsCoreData/ch36p1079peopleGroupsCoreData/PeopleLister.swift


import UIKit
import CoreData

func delay(_ delay:Double, closure:@escaping ()->()) {
    let when = DispatchTime.now() + delay
    DispatchQueue.main.asyncAfter(deadline: when, execute: closure)
}


class PeopleLister: UITableViewController, NSFetchedResultsControllerDelegate, UITextFieldDelegate {
    
    let group : Group
    
    init(group:Group) {
        self.group = group
        super.init(nibName:"PeopleLister", bundle:nil)
    }
    
    required init(coder: NSCoder) {
        fatalError("NSCoding not supported")
    }
    
    lazy var ds : UITableViewDiffableDataSource<String,NSManagedObjectID> = {
        UITableViewDiffableDataSource(tableView: self.tableView) {
            tv,ip,id in
            let cell = tv.dequeueReusableCell(withIdentifier: "Person", for: ip)
            let person = self.frc.object(at: ip)
            let first = cell.viewWithTag(1) as! UITextField
            let last = cell.viewWithTag(2) as! UITextField
            first.text = person.firstName; last.text = person.lastName
            first.delegate = self; last.delegate = self
            return cell
        }
    }()
    
    lazy var frc: NSFetchedResultsController<Person> = {
        let req: NSFetchRequest<Person> = Person.fetchRequest()
        req.fetchBatchSize = 20
        let sortDescriptor = NSSortDescriptor(key:"timestamp", ascending:true)
        req.sortDescriptors = [sortDescriptor]
        
        let pred = NSPredicate(format:"group = %@", self.group)
        req.predicate = pred
        
        let afrc = NSFetchedResultsController(fetchRequest:req,
                                              managedObjectContext:self.group.managedObjectContext!,
                                              sectionNameKeyPath:nil, cacheName:nil)
        afrc.delegate = self
        
        do {
            try afrc.performFetch()
        } catch {
            print("Unresolved error \(error)")
            fatalError("Aborting with unresolved error")
        }
        return afrc
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = self.group.name
        let b = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(doAdd))
        self.navigationItem.rightBarButtonItems = [b]
        
        self.tableView.register(UINib(nibName: "PersonCell", bundle: nil), forCellReuseIdentifier: "Person")
        
        _ = self.frc // "tickle" the lazies
    }
        
    @objc func doAdd(_:AnyObject) {
        self.tableView.endEditing(true)
        let context = self.frc.managedObjectContext
        let person = Person(context:context)
        person.group = self.group
        person.lastName = ""
        person.firstName = ""
        person.timestamp = Date()
        delay(0.1) {
            let row = self.tableView.numberOfRows(inSection: 0) - 1
            let ip = IndexPath(row: row, section: 0)
            self.tableView.scrollToRow(at: ip, at: .bottom, animated: false)
            let cell = self.tableView.cellForRow(at: ip)
            if let cell = cell {
                if let tf = cell.contentView.viewWithTag(1) as? UITextField {
                    tf.becomeFirstResponder()
                }
            }
        }
    }
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        print("did end editing")
        var v : UIView = textField
        repeat { v = v.superview! } while !(v is UITableViewCell)
        let cell = v as! UITableViewCell
        let ip = self.tableView.indexPath(for:cell)!
        let object = self.frc.object(at:ip)
        object.setValue(textField.text!, forKey: ((textField.tag == 1) ? "firstName" : "lastName"))
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.view.endEditing(true)
        (UIApplication.shared.delegate as! AppDelegate).saveContext()
    }
    
    // === content update ===
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChangeContentWith snapshot: NSDiffableDataSourceSnapshotReference) {
        // it's a string and a NSManagedObjectID
        //        var snap = NSDiffableDataSourceSnapshot<String,NSManagedObjectID>()
        //        snap.appendSections(snapshot.sectionIdentifiers as! [String])
        //        for item in snapshot.itemIdentifiers {
        //            let section = snapshot.sectionIdentifier(forSectionContainingItemIdentifier: item)
        //            snap.appendItems([item as! NSManagedObjectID], toSection: section as? String)
        //        }
        //        self.ds.apply(snap, animatingDifferences: false)
        print("change")
        let oldsnapshot = self.ds.snapshot()
        let snapshot = snapshot as NSDiffableDataSourceSnapshot<String,NSManagedObjectID>
        if oldsnapshot.itemIdentifiers == snapshot.itemIdentifiers {
            return // prevent reload in case where nothing happened
        }
        self.ds.apply(snapshot, animatingDifferences: false)
    }

}
